Fixes Loss for TransfoXL when using Trainer API

gugarosa opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

This PR address the problem of using TransfoXL with the Trainer API by introducing straightforward changes, as follows:

- Changes the `losses` parameter from TransfoXL output to `loss`, allowing its usage with the Trainer API;
- Reduces the loss by mean calculation and by ignoring pad tokens;
- Adds a checker that will validate whether input labels are not only composed by pad tokens, which eventually breaks the loss calculation (tested on a full wikitext-103 version). 

I have been working and training TransfoXL from scratch for the past months, and so far this is the most stable code that I could get with huggingface/transformers.

Let me know if there is anything else that should be changed!

PS: fast testing script: `python examples/pytorch/language-modeling/run_clm.py --model_name_or_path transfo-xl-wt103 --dataset_name wikitext --dataset_config_name wikitext-103-raw-v1 --output_dir out`

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [X] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [X] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@patrickvonplaten 

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- albert, bert, xlm: @LysandreJik
- blenderbot, bart, marian, pegasus, encoderdecoder,  t5: @patrickvonplaten, @patil-suraj
- longformer, reformer, transfoxl, xlnet: @patrickvonplaten
- fsmt: @stas00
- funnel: @sgugger
- gpt2: @patrickvonplaten, @LysandreJik
- rag: @patrickvonplaten, @lhoestq
- tensorflow: @LysandreJik

Library:

- benchmarks: @patrickvonplaten
- deepspeed: @stas00
- ray/raytune: @richardliaw, @amogkam
- text generation: @patrickvonplaten
- tokenizers: @n1t0, @LysandreJik
- trainer: @sgugger
- pipelines: @LysandreJik

Documentation: @sgugger

HF projects:

- datasets: [different repo](https://github.com/huggingface/datasets)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Examples:

- maintained examples (not research project or legacy): @sgugger, @patil-suraj
- research_projects/bert-loses-patience: @JetRunner
- research_projects/distillation: @VictorSanh

 -->

🚨🚨 **Breaking change TransfoXL** 🚨🚨

This PR introduces a breaking change.
Previously the output attribute ``losses`` can be found at index 0. Now it is at index 2 in case `labels` are passed. 

This means if you code previously used:

```python
losses = model(input_ids, labels=labels)[0]
```

it should now change to:

```python
losses = model(input_ids, labels=labels)[2]
```

gugarosa committed:
fix(transfo_xl): Fixes TransfoXL support when using Trainer.

gugarosa committed:
fix(tests): Uses losses_1 and losses_2 pattern with TransfoXL test.

patrickvonplaten commented on src/transformers/models/transfo_xl/modeling_transfo_xl.py:
```diff
@@ -673,8 +673,8 @@ class TransfoXLLMHeadModelOutput(ModelOutput):
     Base class for model's outputs that may also contain a past key/values (to speed up sequential decoding).
 
     Args:
-        losses (:obj:`torch.FloatTensor` of shape `(batch_size, sequence_length-1)`, `optional`, returned when ``labels`` is provided)
```
This is a pretty strong backward breaking change. Could we maybe instead try to just alias to losses with a `@property` function, e.g.:

```python
@property
def loss(self):
    return self.losses
```

patrickvonplaten commented on src/transformers/models/transfo_xl/modeling_transfo_xl.py:
```diff
@@ -1097,16 +1097,30 @@ def forward(
+            miss_valid_label = labels[0, 1:].sum() == (labels.size(1) - 1) * -100
+            if miss_valid_label:
+                # Sets an <EOS> token, just to prevent loss from being NaN
+                labels[0, 1] = self.config.eos_token_id
```
Could you give a bit more background on why this is necessary? I don't really see why the exactly those labels should be EOS

gugarosa replied:
Since we are dealing with a customizable softmax/loss layer [ProjectiveAdaptiveLogSoftmax](https://github.com/huggingface/transformers/blob/master/src/transformers/models/transfo_xl/modeling_transfo_xl_utilities.py#L29), it is unable to deal when all of the labels are out-of-vocabulary tokens, for example, pad tokens which were replaced with `-100`. Thus, if we feed a label which is fully composed of `-100`, the output loss would be `0`, its average would be `0` and it would throw an error when trying to be backwarded. 

Ideally, would be nice if we could mimic the behavior of the [CrossEntropyLoss](https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html), which takes into an account a index that can be ignored (`ignore_index=-100`), and is internally able to deal whenever all labels have that particular index.

patrickvonplaten replied:
Thanks!

patrickvonplaten commented on src/transformers/models/transfo_xl/modeling_transfo_xl.py:
```diff
@@ -1097,16 +1097,30 @@ def forward(
+        if labels is not None:
+            loss = softmax_output.view(bsz, tgt_len - 1)
+            # Avoids from incorporating padding (-100) tokens into loss value
+            loss = loss[loss != 0].mean()
```
here we go from a a vector to a scalar which is also very much a breaking change. Why do we need to do this?

gugarosa replied:
I would say that since we are using a customizable loss layer, which is the ProjectiveAdaptiveLogSoftmax (https://github.com/huggingface/transformers/blob/master/src/transformers/models/transfo_xl/modeling_transfo_xl_utilities.py#L29) and it outputs a vector instead of a scalar, we are unable to perform the backward pass inside the Trainer API when it comes to the time (https://github.com/huggingface/transformers/blob/master/src/transformers/trainer.py#L1885).

If we gather GPT2 as an example, its loss is already reduced when it comes out of the CrossEntropyLoss (https://github.com/huggingface/transformers/blob/master/src/transformers/models/gpt2/modeling_gpt2.py#L1075).

patrickvonplaten commented:
Thanks a lot for the very in-detail explanations @gugarosa ! I like your proposals and IMO we could do the following:

1) Output both `loss` and `losses` whereas `losses` is still in "vector"-form and we introduce a new `loss` output that has your proposed form. So the following use cases should work:

```python
from transformers import TransfoXLLMHead

model = TransfoXLLMHead.from_pretrained(...)

# 1) keep backward compatibility -> this should be an output in vector form
losses = model(input_ids).losses 

# 2) allow model to be used in trainer -> this should be in scalar form
loss = model(input_ids).loss 
```

2) Then regarding the tuple order we will probably have to do a breaking change with the new tuple order looking as follows:
(loss, ...., lossses)   <- this means that 

```python
losses = model(input_ids)[0] 
# is now changed to
loss = model(input_ids)[0] 
```

which is ok IMO given that most people use `outputs.losses` ...

What do you think @sgugger @LysandreJik ?

sgugger reviewed (commented):
This is very breaking as @patrickvonplaten mentioned, but I think we can easily make it less breaking by adapting this PR to:
- keep `losses` in the outputs as a tensor and move it after the logits (that would be breaking for people who indexed into the output, but with an easy way to fix, just change that index)
- add the new loss computed as a scalar in a loss field (first in the output)

Thanks a lot for fixing this model to work with the `Trainer`!

**Edit** I had not viewed your last comment @patrickvonplaten and I agree with everything you said :-)

patrickvonplaten commented:
Awesome! @gugarosa - feel free to go ahead with the discussed approach if you want :-)

gugarosa committed:
fix(transfo_xl): Adds requested changes to allow for backward compatibility.

fix(transfo_xl): Adds requested changes to allow for backward compatibility.

fix(transfo_xl): Fixes code styling.

gugarosa commented:
Sorry for taking too long, I had a family issue and went off for the week.

Nevertheless, I have updated the PR with the requested changes (thanks so much!), so please feel free to criticize them or request for newer changes. I have also tested with the proposed snippet and everything seems to be working!

```
import torch
from transformers import TransfoXLLMHeadModel

# Constants
BATCH_SIZE = 1
SEQ_LEN = 32

# Loads model from Hub and defines random inputs
model = TransfoXLLMHeadModel.from_pretrained('transfo-xl-wt103')
input_ids = torch.randint(0, 10000, (BATCH_SIZE, SEQ_LEN))

# 1) keep backward compatibility -> this should be an output in vector form
losses = model(input_ids, labels=input_ids).losses 

# 2) allow model to be used in trainer -> this should be in scalar form
loss = model(input_ids, labels=input_ids).loss 

# Asserts the outputs' shapes
assert losses.shape == (BATCH_SIZE, SEQ_LEN - 1)
assert loss.shape == ()
```

Best regards,
Gustavo.

sgugger reviewed (approved):
Thanks for amending your PR! As there is a slight breaking change, making sure @LysandreJik is okay as well.

Also make sure to run `make style` on your branch to fix the code quality issue :-)

gugarosa commented:
Thanks for the feedback! I have squashed the commit with the styling changes and all tests seems to be passing now.

patrickvonplaten commented:
Cool looks good to me! Left a 🚨🚨 breaking 🚨🚨 message at the top of the PR to make users aware

patrickvonplaten commented:
@LysandreJik - feel free to merge if you're ok with the small breaking change

patrickvonplaten commented:
@LysandreJik - that sounds good to me. Think it does make sense to aim for 100% backward compatibility here. @sgugger what do you think?

sgugger commented:
That works for me. Also wants to highlight that this would only be necessary if the user specifies `return_dict=False`, as when the outputs is one of our `ModelOutput`, the `Trainer` indexes with the key "loss" and not the first index.

LysandreJik commented:
@gugarosa, would you like to take a stab at this? Otherwise let me know and I'm happy to help out :)

LysandreJik commented:
I'll try and make time for this this week if nobody beats me to it.

LysandreJik commented:
Couldn't push directly to your fork so I opened a PR here: https://github.com/gugarosa/transformers/pull/1

LysandreJik commented:
@gugarosa please let me know if you'd like me to take over the PR. I believe all that remains is to merge the PR I opened here: https://github.com/gugarosa/transformers/pull/1 and we can merge this current PR then.

Thanks!
